ocaml/testsuite/tests/typing-layouts-err-msg/value.ml

```diff
@@ -0,0 +1,433 @@
+Error: The layout of type attr is value, because
+         it's an extensible variant.
+       But the layout of type attr must be a sublayout of void, because
+         of the annotation on the declaration of the type attr.
+|}]
+
+(* Primitive *)
+let f : unit -> ('a : void) = fun () -> "abc"
+[%%expect{|
+Line 1, characters 40-45:
+1 | let f : unit -> ('a : void) = fun () -> "abc"
+                                            ^^^^^
+Error: This expression has type string but an expression was expected of type
+         ('a : void)
+       The layout of string is value, because
+         it is the primitive value type string.
+       But the layout of string must be a sublayout of void, because
+         of the annotation on the type variable 'a.
+|}];;
+
+(* Type_argument *)
+let f (x: t_void) = [x]
+[%%expect{|
+Line 1, characters 21-22:
+1 | let f (x: t_void) = [x]
+                         ^
+Error: This expression has type t_void but an expression was expected of type
+         ('a : value)
+       The layout of t_void is void, because
+         of the definition of t_void at line 6, characters 0-19.
+       But the layout of t_void must be a sublayout of value, because
+         the type argument of list has layout value.
+|}];;
+
+(* Tuple *)
+let f : unit -> ('a : void) = fun () -> (1,2)
+[%%expect{|
+Line 1, characters 40-45:
+1 | let f : unit -> ('a : void) = fun () -> (1,2)
+                                            ^^^^^
+Error: This expression has type 'b * 'c
+       but an expression was expected of type ('a : void)
+       The layout of 'a * 'b is value, because
+         it's a tuple type.
+       But the layout of 'a * 'b must be a sublayout of void, because
+         of the annotation on the type variable 'a.
+|}];;
+
+(* Row_variable *)
+(* CR layouts v2.9: add test *)
+
+(* Polymorphic_variant *)
+type ('a: void) t = 'a
+let f (x: [`A of int | `B]): 'a t = x
+[%%expect{|
+type ('a : void) t = 'a
+Line 2, characters 36-37:
+2 | let f (x: [`A of int | `B]): 'a t = x
+                                        ^
+Error: This expression has type [ `A of int | `B ]
+       but an expression was expected of type 'a t = ('a : void)
+       The layout of [ `A of int | `B ] is value, because
+         it's a polymorphic variant.
+       But the layout of [ `A of int | `B ] must be a sublayout of void, because
+         of the definition of t at line 1, characters 0-22.
+|}]
+
+(* Arrow *)
+type ('a: void) t = 'a
+let f (x : int -> int): 'a t = x
+[%%expect{|
+type ('a : void) t = 'a
+Line 2, characters 31-32:
+2 | let f (x : int -> int): 'a t = x
+                                   ^
+Error: This expression has type int -> int
+       but an expression was expected of type 'a t = ('a : void)
+       The layout of int -> int is value, because
+         it's a function type.
+       But the layout of int -> int must be a sublayout of void, because
+         of the definition of t at line 1, characters 0-22.
+|}]
+
+(* Tfield *)
```

Comment:
`Tfield` and `Tnil` are very strange. As the message suggests, they shouldn't actually happen. Maybe just clarify the text to call for help, but we don't need to add tests here.